### PR TITLE
Encode all URI components…

### PR DESCRIPTION
**Pull request by mickaeltr**
… except for the host name, because url.parse does not seem to handle that:

```
> url.parse("http://cänal/")
Url {
  protocol: 'http:',
  slashes: true,
  auth: null,
  host: 'xn--cnal-loa',
  port: null,
  hostname: 'xn--cnal-loa',
  hash: null,
  search: null,
  query: null,
  pathname: '/',
  path: '/',
  href: 'http://xn--cnal-loa/' }

> url.parse(encodeURI("http://cänal/"))
Url {
  protocol: 'http:',
  slashes: true,
  auth: null,
  host: 'c',
  port: null,
  hostname: 'c',
  hash: null,
  search: null,
  query: null,
  pathname: '%C3%A4nal/',
  path: '%C3%A4nal/',
  href: 'http://c/%C3%A4nal/' }
```

### CODE REVIEW

**Comment by mickaeltr** [comment]
C'est de la merde ? Quoi ?

**Comment by azu** [comment]
Any update?

**Comment by mickaeltr** [comment]
@azu Nope… so I forked it: https://github.com/mickaeltr/safe-url-assembler

**Comment by azu** [comment]
@mickaeltr I see. Thanks.

File: lib/url-assembler-factory.js [hunk 1]
```diff
@@ -83,6 +83,8 @@ module.exports = function (request) {
     if (typeof key === 'object') {
       return _multiParam(this, key, (value === true));
     }
+    key = encodeURI(key);
+    value = encodeURI(value);
```
Comment (azu) [inline lib/url-assembler-factory.js]: `/` param still not encoded. 
[encodeURI()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI "encodeURI()") does't encode `/`.  [encodeURIComponent()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent "encodeURIComponent()") does.

@mickaeltr Is this expected behavior?

```js
const url = UrlAssembler('http://example.com')
    .template('search/:q')
    .param({
        q: "/"
    }).toString();
console.log(url); // http://example.com//
```

Comment (mickaeltr) [inline lib/url-assembler-factory.js]: @azu That does not seem right. Feel free to make a pull request to safe-url-assembler and I'll release a new version.

Comment (azu) [inline lib/url-assembler-factory.js]: Ok. I've submitted PR.
https://github.com/mickaeltr/safe-url-assembler/pull/2

**Comment by Floby** [comment]
Fixed in latest version.

**Pull request closed without merging**